examples/batched-solver-from-files/batched-solver-from-files.cpp

```diff
@@ -0,0 +1,223 @@
+        std::ifstream mtx_fd(fname);
+        std::string b_str = "b.mtx";
+        data[i] = gko::read_raw<value_type>(mtx_fd);
+        std::string bfname =
+            "data/" + problem_descr_str + "/" + std::to_string(i) + "/" + b_str;
+        std::ifstream b_fd(bfname);
+        bdata[i] = gko::read_raw<value_type>(b_fd);
+        // If necessary, 'scaling vectors' can be provided to diagonal-scale
+        //  a system from the left and the right. For this example, no
+        //  scaling vectors are provided.
+        if (batch_scaling == "explicit") {
+            std::string scale_fname = fbase + "S.mtx";
+            std::ifstream scale_fd(scale_fname);
+            scale_data[i] = gko::read_raw<value_type>(scale_fd);
+        }
+    }
+    auto single_batch = mtx_type::create(exec);
+    single_batch->read(data);
+    // We can duplicate the batch a few times if we wish.
+    std::shared_ptr<mtx_type> A =
+        mtx_type::create(exec, num_duplications, single_batch.get());
+    // Create RHS
+    auto temp_b = vec_type::create(exec);
+    temp_b->read(bdata);
+    auto b = vec_type::create(exec, num_duplications, temp_b.get());
+    // Create initial guess as 0 and copy to device
+    const size_type num_total_systems = num_systems * num_duplications;
+    auto x = vec_type::create(exec);
+    auto host_x = vec_type::create(exec->get_master(), b->get_size());
+    // The number of rows in each system is taken as the 0th dimension of the
+    //  size of the 0th system in the batch.
+    const int num_rows = static_cast<int>(b->get_size().at(0)[0]);
+    for (size_type isys = 0; isys < num_total_systems; isys++) {
+        for (int irow = 0; irow < num_rows; irow++) {
+            host_x->at(isys, irow, 0) = gko::zero<value_type>();
+        }
+    }
+    x->copy_from(host_x.get());
+
+    const real_type reduction_factor{1e-6};
+
+    // @sect3{Create the batch solver factory}
+    // Create a batched solver factory with relevant parameters.
+    auto solver_gen =
+        solver_type::build()
+            .with_max_iterations(500)
+            .with_residual_tol(reduction_factor)
+            .with_tolerance_type(gko::stop::batch::ToleranceType::relative)
+            .with_preconditioner(gko::preconditioner::batch::type::jacobi)
+            .on(exec);
+
+    // @sect3{Batch logger}
+    // Create a logger to obtain the iteration counts and "implicit" residual
+    //  norms for every system after the solve.
+    std::shared_ptr<const gko::log::BatchConvergence<value_type>> logger =
+        gko::log::BatchConvergence<value_type>::create(exec);
+
+    // Generate the batch solver from the batch matrix
+    auto solver = solver_gen->generate(A);
+
+    // add the logger to the solver
+    solver->add_logger(logger);
+    // Solve the batch system
+    solver->apply(lend(b), lend(x));
+    // This is not necessary, but one might want to remove the logger before
+    //  the next solve using the same solver object.
```

Comment:
Can we maybe add a simple chrono timing here to make it easier to time this ?

Reply:
Agree here. The issue is maybe since it's batched you want a GPU-side timer for better resolution? At least a CPU-side timer would be useful though.